In templates, autocomplete string constants, number constants, `true`, `false`, `null`, and `undefined`

zarend opened the pull request:
# fix(compiler-cli): autocomplete literal types in templates. …

This adds string literals, number literals, `true`, `false`, `null` and
`undefined` to autocomplete results in templates.

For example, when completing an input of union type.

Component: `@Input('input') input!: 'a'|'b'|null;`
Template: `[input]="|"`

Provide `'a'`, `'b'`, and `null` as autocompletion entries.

Previously we did not include literal types because we only included
results from the component context (`ctx.`) and the template scope.

## out-of-scope
Completing method calls with no parameters is out-of-scope. This is a difficult problem for a few reasons.

directive class: `foo(x: 1|2|3) {}`
template: `{{foo()}}`

The problem is that this maps to a `MethodCall` AST node with no arguments. From looking at the AST, we have no way to know if the cursor is inside the parens, or on the name of the function. We don't know if it is `f|oo()` or `foo(|)`.

Also, we cannot insert `undefined` as a parameter. Even though arguments not passed are implicitly undefined. This causes a type error if the function signature takes no arguments.

zarend committed:
fix(compiler-cli): autocomplete literal types in templates.

This adds string literals, number literals, `true`, `false`, `null` and
`undefined` to autocomplete results in templates.

For example, when completing an input of union type.

Component: `@Input('input') input!: 'a'|'b'|null;`
Template: `[input]="|"`

Provide `'a'`, `'b'`, and `null` as autocompletion entries.

Previously we did not include literal types because we only included
results from the component context (`ctx.`) and the template scope.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -38,10 +38,12 @@ export class CompletionEngine {
-    }
+  getGlobalCompletions(context: TmplAstTemplate|null, node: AST|TmplAstNode): GlobalCompletion
+      |null {
+    // TODO(zarend): re-enable the cache
```
I think for the cache, we should focus on caching the stuff that doesn't change (`templateContext` and `componentContext`).

zarend replied:
Done ✔️

atscott commented on packages/language-service/ivy/completions.ts:
```diff
@@ -638,6 +656,23 @@ export class CompletionBuilder<N extends TmplAstNode|AST> {
     };
   }
+
+  private isAstCompletion(completion: ts.CompletionEntry): boolean {
```
I'm not sold on this name. What does it mean to be an AST completion? When I saw this above, I had no idea what it could possibly mean and had to come look at the implementation.

zarend replied:
okay, I renamed the completions from the AST node of the cursor to `nodeCompletions`, and renamed this function to `isValidNodeContextCompletion`.

atscott commented on packages/language-service/ivy/completions.ts:
```diff
@@ -638,6 +656,23 @@ export class CompletionBuilder<N extends TmplAstNode|AST> {
   }
+
+  private isAstCompletion(completion: ts.CompletionEntry): boolean {
+    // include completion of string literals, number literals, `true`, `false`, `null`, and
+    // `undefined`
```
This should be a javadoc rather than a comment at the top of the function.

atscott commented on packages/language-service/ivy/completions.ts:
```diff
@@ -244,6 +244,24 @@ export class CompletionBuilder<N extends TmplAstNode|AST> {
 
+    // Merge TS completion results with results from the ast context.
+    if (astContext !== null) {
+      const tsLsAstContextCompletions = this.tsLS.getCompletionsAtPosition(
```
optional nit: just `tsCompletions` or `completions` for the variable name would be sufficient imo. You _did_ simplify it to `tsCompletion` below so it seems like maybe you agree :)

atscott reviewed (approved):
Looks good overall! 👍

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -25,65 +25,49 @@ import {TemplateData} from './context';
+   * (`null`). Memoizes `getContextCompletions`.
    */
-  private globalCompletionCache = new Map<TmplAstTemplate|null, GlobalCompletion>();
+  private contextCompletionCache = new Map<TmplAstTemplate|null, GlobalCompletion>();
```
In general this shouldn't store a type that includes more information than we want to cache. Using a narrower type here would've prevented the bug below.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -25,65 +25,49 @@ import {TemplateData} from './context';
+      const astLocation =
+          findFirstMatchingNode(this.tcb, {filter: ts.isIdentifier, withSpan: node.sourceSpan});
+      if (astLocation !== null) {
+        completion.astContext = {
```
This is mutating a value that's stored in the `contextCompletionCache`, polluting the cache and potentially corrupting other responses.

We should avoid doing this by:
* defining a separate type for the cache that doesn't have the `ast` context.
* copying the results of `getContextCompletion()` into a _new_ object of type `GlobalCompletion` (which we'd be required to do as the type of the old one wouldn't have the field we need.

zarend replied:
🤦 good catch

zarend commented:
Hi @alxhub, I made fixes and this is ready for your eyes again 👀 . Since the last time you reviewed this, I

- renamed the AST node of the cursor position to the `nodeContext`
- Fixed the cache pollution problem in `CompletionEngine` by decoupling the component context from the template context – then storing only the template context in the cache.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -23,65 +23,69 @@ import {TemplateData} from './context';
-
-    const completion: GlobalCompletion = {
-      componentContext: {
+      this.componentContext = null;
```
Generally, put the common case (`globalRead !== null`) in the "then" branch of the `if`.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -23,65 +23,69 @@ import {TemplateData} from './context';
 
-    this.globalCompletionCache.set(context, completion);
-    return completion;
+    return {componentContext: this.componentContext, templateContext, nodeContext: astContext};
```
Use a trailing `,` after `astContext` so the formatter will space this out nicely.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -23,65 +23,69 @@ import {TemplateData} from './context';
+      return null;
+    }
+    const templateContext = this.getTemplateContextCompletions(context);
+
```
Nit: put this whitespace above, separating the section that deals with `templateContext`.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/completion.ts:
```diff
@@ -23,65 +23,69 @@ import {TemplateData} from './context';
+      const astLocation =
+          findFirstMatchingNode(this.tcb, {filter: ts.isIdentifier, withSpan: node.sourceSpan});
+      if (astLocation !== null) {
+        astContext = {shimPath: this.shimPath, positionInShimFile: astLocation.getStart()};
```
Same here - a trailing comma will cause this to format much more nicely.
